## Reading files

In [4]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [5]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [6]:
wh = pd.concat([wh2018,wh2019])

In [7]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [18]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

Para encontrar los datos nulos, primero verifique la información de cada columna. 

In [50]:
Df= pd.DataFrame(wh)

info_DF = Df.info()




<class 'pandas.core.frame.DataFrame'>
Int64Index: 312 entries, 0 to 155
Data columns (total 10 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  312 non-null    int64  
 1   Country or region             312 non-null    object 
 2   Score                         293 non-null    float64
 3   GDP per capita                300 non-null    float64
 4   Social support                296 non-null    float64
 5   Healthy life expectancy       286 non-null    float64
 6   Freedom to make life choices  289 non-null    float64
 7   Generosity                    291 non-null    float64
 8   Perceptions of corruption     299 non-null    float64
 9   year                          312 non-null    int64  
dtypes: float64(7), int64(2), object(1)
memory usage: 26.8+ KB


Luego mostré con el .head los valores falsos y verdaderos, donde los verdaderos suman como nulos. 

In [52]:
valoresNulos = Df.isna()
valoresNulos.head()

Overall rank  Country or region  Score  GDP per capita  Social support  \
0         False              False  False           False           False   
1         False              False  False           False           False   
2         False              False  False           False           False   
3         False              False  False           False            True   
4         False              False  False           False           False   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    False                         False       False   
1                    False                         False       False   
2                    False                         False       False   
3                    False                         False        True   
4                    False                         False       False   

   Perceptions of corruption   year  
0                      False  False  
1                      False  False  
2                      False  False  
3                      False  False  
4                      False  False

Aquí sumo todos lo valores por columna individual

In [53]:
valoresNulosTotales = valoresNulos.sum()
valoresNulosTotales 

Overall rank                     0
Country or region                0
Score                           19
GDP per capita                  12
Social support                  16
Healthy life expectancy         26
Freedom to make life choices    23
Generosity                      21
Perceptions of corruption       13
year                             0
dtype: int64

Luego lo convertí a porcentaje

In [58]:
valoresNulosTotalesPorcentaje = (valoresNulosTotales/100)
valoresNulosTotalesPorcentaje

Overall rank                    0.00
Country or region               0.00
Score                           0.19
GDP per capita                  0.12
Social support                  0.16
Healthy life expectancy         0.26
Freedom to make life choices    0.23
Generosity                      0.21
Perceptions of corruption       0.13
year                            0.00
dtype: float64

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

Realice un groupby he hice un conteo para conocer la cantidad de datos en cada indice.

In [71]:
resumen = wh.groupby(["year"], as_index=False).agg({"Score": ["count"] , "Social support": ["count"] , "Overall rank": ["count"] , "Country or region": ["count"], "GDP per capita": ["count"], "Healthy life expectancy": ["count"],
       "Freedom to make life choices": ["count"], "Generosity": ["count"], "Perceptions of corruption": ["count"] } )
resumen

year Score Social support Overall rank Country or region GDP per capita  \
        count          count        count             count          count   
0  2018   151            141          156               156            147   
1  2019   142            155          156               156            153   

  Healthy life expectancy Freedom to make life choices Generosity  \
                    count                        count      count   
0                     144                          144        139   
1                     142                          145        152   

  Perceptions of corruption  
                      count  
0                       149  
1                       150

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

<bound method DataFrame.count of      Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.592                    0.874    

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro